src/realm_dart_logger.cpp

```diff
@@ -0,0 +1,113 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+#include <sstream>
+#include <set>
+#include <mutex>
+#include <thread>
+#include <map>
+#include <chrono>
+
+#include "realm_dart_logger.h"
+#include <realm/object-store/c_api/types.hpp>
+
+auto& dart_logger_mutex = *new std::mutex;
+bool is_core_logger_callback_set = false;
+std::map<Dart_Port, realm_log_level_e> dart_send_ports;
+realm_log_level_e default_log_level = RLM_LOG_LEVEL_INFO;
+
+
+bool isPortRegistered(Dart_Port key)
+{
+    return (dart_send_ports.find(key) != dart_send_ports.end());
+}
+
+realm_log_level_e calucale_minimum_log_level()
+{
+    auto min_log_level = default_log_level;
+    if (dart_send_ports.size() > 0)
+    {
+        auto min_element = std::min_element(dart_send_ports.begin(), dart_send_ports.end(),
+            [](std::pair<Dart_Port, realm_log_level_e> const& prev, std::pair<Dart_Port, realm_log_level_e> const& next) {
+            return prev.second < next.second;
+        });
+        min_log_level = min_element->second;
+    }
+    return min_log_level;
+}
+
+RLM_API void realm_dart_release_logger(Dart_Port port) {
+    std::lock_guard lock(dart_logger_mutex);
+    if (isPortRegistered(port))
+    {
+        dart_send_ports.erase(port);
+        auto minimum_level = calucale_minimum_log_level();
+        realm_set_log_level(minimum_level);
+    }
+}
+
+bool send_message_to_scheduler(Dart_Port port, realm_log_level_e level, const char* message)
+{
+    Dart_CObject c_level;
+    c_level.type = Dart_CObject_kInt32;
+    c_level.value.as_int32 = level;
+
+    Dart_CObject c_message;
+    c_message.type = Dart_CObject_kString;
+    c_message.value.as_string = (char*)message;
+
+    Dart_CObject* c_request_arr[] = { &c_level , &c_message };
+    Dart_CObject c_request;
+    c_request.type = Dart_CObject_kArray;
+    c_request.value.as_array.values = c_request_arr;
+    c_request.value.as_array.length = sizeof(c_request_arr) / sizeof(c_request_arr[0]);
+
+    bool result = Dart_PostCObject_DL(port, &c_request);
+    return result;
+}
+
+void realm_dart_logger_callback(realm_userdata_t userData, realm_log_level_e level, const char* message) {
+    std::lock_guard lock(dart_logger_mutex);
+
+    for (auto itr = dart_send_ports.begin(); itr != dart_send_ports.end(); ++itr) {
+        Dart_Port port = itr->first;
+        bool result = send_message_to_scheduler(port, level, message);
+    }
+}
+
+RLM_API bool realm_dart_init_default_logger() {
+    std::lock_guard lock(dart_logger_mutex);
+    if (is_core_logger_callback_set) {
+        return false;
+    }
+    realm_set_log_callback(realm_dart_logger_callback, default_log_level, nullptr, nullptr);
+    is_core_logger_callback_set = true;
+    return true;
+}
+
+RLM_API void realm_dart_set_log_level(realm_log_level_e level, Dart_Port port)
+{
+    std::lock_guard lock(dart_logger_mutex);
+    if (dart_send_ports[port] != level)
```

Comment:
If the port does not exists?

Reply:
If there is no item for this port in the map, an item will be automatically added even if you try to get it `dart_send_ports[port]`.

Reply:
Ok, checked with `dart_send_ports.find(port)`